include/utils/float_comparison.hpp

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2019 Codeplay Software Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use these files except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * This file provides tools to compare floating-point numbers.
+ * The function almost_equal returns a boolean indicating if two scalars can
+ * be considered equal. The function compare_vectors checks if two vectors of
+ * the same size are almost equal, prints a message on stderr if a value
+ * mismatch is found, and returns a boolean as well.
+ * Neither of the methods raises an exception nor calls exit(1)
+ */
+
+#ifndef UTILS_FLOAT_COMPARISON_H_
+#define UTILS_FLOAT_COMPARISON_H_
+
+#include <cmath>
+
+namespace utils {
+
+/**
+ * Indicates the tolerated margin for relative differences
+ */
+template <typename scalar_t>
+inline scalar_t getRelativeErrorMargin() {
+  return static_cast<scalar_t>(0.005);
+}
+
+template <>
+inline double getRelativeErrorMargin<double>() {
+  return 0.000001;
+}
+
+/**
+ * Indicates the tolerated margin for absolute differences
```

Comment:
Put a comment and explain how did you came up with these numbers?

Reply:
I've added a few comments

Reply:
I have added even more explanations. I hope it's clear.
Also I've made a change to properly handle the case of zero